go/stellar/acctbundle/acctbundle.go

```diff
@@ -0,0 +1,683 @@
+	if err != nil {
+		return empty, err
+	}
+	visibleHash := sha256.Sum256(visiblePack)
+	secretV2 := versioned.V2()
+	if !hmac.Equal(visibleHash[:], secretV2.VisibleHash) {
+		return empty, errors.New("corrupted bundle: visible hash mismatch")
+	}
+	var visibleV2 stellar1.BundleVisibleV2
+	err = libkb.MsgpackDecode(&visibleV2, visiblePack)
+	if err != nil {
+		return empty, err
+	}
+	return merge(secretV2, visibleV2)
+}
+
+// decryptParent decrypts an encrypted parent bundle with the provided puk.
+func decryptParent(encBundle stellar1.EncryptedBundle, puk libkb.PerUserKeySeed) (stellar1.BundleSecretVersioned, error) {
+	var empty stellar1.BundleSecretVersioned
+	if encBundle.V != 2 {
+		return empty, fmt.Errorf("invalid stellar secret account bundle encryption version: %v", encBundle.V)
+	}
+
+	// Derive key
+	reason := libkb.DeriveReasonPUKStellarBundle
+	symmetricKey, err := puk.DeriveSymmetricKey(reason)
+	if err != nil {
+		return empty, err
+	}
+
+	// Secretbox
+	clearpack, ok := secretbox.Open(nil, encBundle.E,
+		(*[libkb.NaclDHNonceSize]byte)(&encBundle.N),
+		(*[libkb.NaclSecretBoxKeySize]byte)(&symmetricKey))
+	if !ok {
+		return empty, errors.New("stellar bundle secret box open failed")
+	}
+
+	// Msgpack (inner)
+	var bver stellar1.BundleSecretVersioned
+	err = libkb.MsgpackDecode(&bver, clearpack)
+	if err != nil {
+		return empty, err
+	}
+	return bver, nil
+}
+
+// decode decodes a base64-encoded encrypted account bundle.
+func decode(encryptedB64 string) (stellar1.EncryptedAccountBundle, stellar1.Hash, error) {
+	cipherpack, err := base64.StdEncoding.DecodeString(encryptedB64)
+	if err != nil {
+		return stellar1.EncryptedAccountBundle{}, stellar1.Hash{}, err
+	}
+	encHash := sha256.Sum256(cipherpack)
+	var enc stellar1.EncryptedAccountBundle
+	if err = libkb.MsgpackDecode(&enc, cipherpack); err != nil {
+		return stellar1.EncryptedAccountBundle{}, stellar1.Hash{}, err
+	}
+	return enc, encHash[:], nil
+}
+
+// unbox unboxes an encrypted account bundle and decodes the visual portion of the bundle.
+// It validates the visible hash in the secret portion.
+func unbox(encBundle stellar1.EncryptedAccountBundle, hash stellar1.Hash /* visB64 string, */, puk libkb.PerUserKeySeed) (*stellar1.AccountBundle, stellar1.AccountBundleVersion, error) {
+	versioned, err := decrypt(encBundle, puk)
+	if err != nil {
+		return nil, 0, err
+	}
+	version, err := versioned.Version()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	var bundleOut stellar1.AccountBundle
+	switch version {
+	case stellar1.AccountBundleVersion_V1:
+		secretV1 := versioned.V1()
+		bundleOut = stellar1.AccountBundle{Signers: secretV1.Signers}
+	case stellar1.AccountBundleVersion_V2,
+		stellar1.AccountBundleVersion_V3,
+		stellar1.AccountBundleVersion_V4,
+		stellar1.AccountBundleVersion_V5,
+		stellar1.AccountBundleVersion_V6,
+		stellar1.AccountBundleVersion_V7,
+		stellar1.AccountBundleVersion_V8,
+		stellar1.AccountBundleVersion_V9,
+		stellar1.AccountBundleVersion_V10:
```

Comment:
What are all these future version numbers for? There's no forward-compatible decoder for bundles

Reply:
They give an opportunity to give a different error when a newer version comes in to an old client.